### ISSUE TITLE
Embeddings for large KG

### ISSUE THREAD

**Issue by deeplifyde** [IBCNServices/pyRDF2Vec#154]
## ❓ Question

Hello,
I'm currently trying to generate Embeddings for a graph with ~ 13 million entities and 15 million walks. I'm using a machine with 128 Gb RAM, however, the in-memory saved random walks overflow the memory. Is there a way to store them to disk and batch load them like e.g. in image processing pipelines?

**Comment by GillesVandewiele** [comment]
Yes, most definitely! Within the fit() function, the extract_walks() function is called that returns a list of walks that are fed to Word2Vec later on.

https://github.com/IBCNServices/pyRDF2Vec/blob/main/pyrdf2vec/rdf2vec.py#L107

You could call this function on different chunks of entities to do this iteratively. We also have some mechanisms to speed up extraction and/or reduce memory usage:

* Hashing of urls (less bytes needed)
* Using HDT to speed up KG loading time, or even better, directly querying a locally hosted SPARQL endpoint (no loading time)
* Extracting a limited amount of walks per entity, but this reduces accuracy later on
* Excluding certain types of walks (forbidden predicates) of which you know they do not carry much information
* Etc.

However, in the end, you will need to feed all these walks AT ONCE to a Word2Vec or other word embedding network, which could be the main bottleneck...

**Comment by deeplifyde** [comment]
Thank you for the fast answer!

Using the is_update=True on chunks of the data is not possible ? Because the underlying model will store the walks for the old entities too ?

**Comment by GillesVandewiele** [comment]
It's not ideal, as Word2Vec doesn't support iterative updating that well (it is possible but suboptimal). I think the custom data loader will provide better results!
